### PR TITLE
Handling Serialization Errors

### PR DESCRIPTION
**Pull request by mrvaruntandon**
Pull request checklist
----

- [X] Please read our [contributor guide](https://github.com/Netflix/dgs-framework/blob/master/CONTRIBUTING.md)
- [ ] Consider creating a discussion on the [discussion forum](https://github.com/Netflix/dgs-framework/discussions)
  first
- [X] Make sure the PR doesn't introduce backward compatibility issues
- [X] Make sure to have sufficient test cases

Pull Request type
----

- [X] Bugfix
- [ ] Feature
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] Other (please describe):

Changes in this PR
----
- Introduced a ErrorDetail Type for Serialization Errors
- Added the new ErrorDetail Type while `TypedGraphQLError` is being formed in `GraphQLJavaErrorInstrumentation` 

_Describe the new behavior from this PR, and why it's needed_
This PR resolves https://github.com/Netflix/dgs-framework/issues/1914

Alternatives considered
----

_Describe alternative implementation you have considered_

### CODE REVIEW

**Review by Meemaw — COMMENTED** [review]
I would argue coercing errors (which the original issue is about) are client errors (not internal server errors).

**Comment by mrvaruntandon** [comment]
Hi @Meemaw, 
> I would argue coercing errors (which the original issue is about) are client errors (not internal server errors).

I checked the `Coercing` Interface, there are three `Coercing*Exception`, one for each method in the interface. I could only come up with scenarios where `parseLiteral()` and `serialize()` in the interface would be invoked to throw their respective exceptions and this PR is based on that understanding.  

`Coercing.parseLiteral()` invoked to resolve an input from a query variable into a Java object, if such parsing fails `CoercingParseLiteralException` is thrown. I tried to replicate this scenario and found my implementation would throw a client error like below:
```json
{
  "errors": [
    {
      "message": "Validation error (WrongType@[ip]) : argument 'domain' with value 'StringValue{value='google'}' is not a valid 'Domain' - Invalid domain name",
      "locations": [
        {
          "line": 11,
          "column": 6
        }
      ],
      "path": [
        "ip"
      ],
      "extensions": {
        "classification": "ValidationError",
        "errorType": "BAD_REQUEST",
        "errorDetail": "INVALID_ARGUMENT"
      }
    }
  ]
}
```
This is not an `INTERNAL` errorType, but a client error type `BAD_REQUEST`. 

The `INTERNAL` errorType would be returned in cases where the server fails to successfully invoke the `serialize()` and throws `CoercingSerializeException`.

Please let me know you are referring to which scenario about coercing errors being client errors and I could test for it and handle it appropriately.

**Comment by mrvaruntandon** [comment]
IMO in summary, `CoercingParseLiteralException` should be treated as client errors and `CoercingSerializeException` should be treated as server errors.

**Comment by mrvaruntandon** [comment]
@paulbakker @Emily @kilink @srinivasankavitha could one of you please review this PR?

**Comment by srinivasankavitha** [comment]
Thanks for the fix! PR changes look good to me.

**Review by srinivasankavitha — APPROVED** [review]

File: graphql-dgs/src/test/kotlin/com/netflix/graphql/dgs/GraphQLJavaErrorInstrumentationTest.kt [hunk 1]
```diff
@@ -157,12 +164,49 @@ class GraphQLJavaErrorInstrumentationTest {
...
+        val result = graphQL.execute(
+            """
+            {
+                ip
+            }
+            """.trimIndent()
+        )
+
+        Assertions.assertThat(result.errors.size).isEqualTo(1)
+        Assertions.assertThat(result.errors[0].extensions.keys.containsAll(listOf("errorDetail", "errorType")))
```
Comment (kilink) [inline graphql-dgs/src/test/kotlin/com/netflix/graphql/dgs/GraphQLJavaErrorInstrumentationTest.kt:179]: This line isn't asserting anything.

Comment (mrvaruntandon) [reply graphql-dgs/src/test/kotlin/com/netflix/graphql/dgs/GraphQLJavaErrorInstrumentationTest.kt]: Hi @kilink , I have added `isTrue` assertion in this test.  At the same time, I see that similar assertion was missing in other tests also, so I have added the assertion in them as well.

**Review by srinivasankavitha — APPROVED** [review]

**Review by Meemaw — APPROVED** [review]

**Comment by mrvaruntandon** [comment]
Hi @srinivasankavitha, could you please tell me why the build associated with this PR is failing? Do I need to take any action to fix the build?

**Comment by kilink** [comment]
@mrvaruntandon it looks like it's failing due to ktlint errors, you can run `./gradlew formatKotlin` to fix them.

**Comment by mrvaruntandon** [comment]
@kilink, thank you for your guidance!

 I have fixed the lint issue, could you please trigger the build again?

**Comment by mrvaruntandon** [comment]
Thank you @srinivasankavitha @kilink @Meemaw for your support. This was my first open source contribution.

**Pull request merged**